lib/her/model/attributes.rb

```diff
@@ -177,42 +144,92 @@ def attribute?(attribute)
       end
 
       module ClassMethods
+
+        # Initialize a single resource
+        #
+        # @private
+        def instantiate_record(klass, parsed_data)
+          if record = parsed_data[:data] and record.kind_of?(klass)
+            record
+          else
+            attributes = klass.parse(record).merge(_metadata: parsed_data[:metadata],
+                                                   _errors: parsed_data[:errors])
+            klass.new(attributes).tap do |record|
+              record.run_callbacks :find
+            end
+          end
+        end
+
+        # Initialize a collection of resources
+        #
+        # @private
+        def instantiate_collection(klass, parsed_data = {})
+          items = klass.extract_array(parsed_data).map do |item|
+            instantiate_record(klass, data: item)
+          end
+          Her::Collection.new(items, parsed_data[:metadata], parsed_data[:errors])
+        end
+
         # Initialize a collection of resources with raw data from an HTTP request
         #
         # @param [Array] parsed_data
         # @private
         def new_collection(parsed_data)
-          Her::Model::Attributes.initialize_collection(self, parsed_data)
+          instantiate_collection(self, parsed_data)
         end
 
         # Initialize a new object with the "raw" parsed_data from the parsing middleware
         #
         # @private
         def new_from_parsed_data(parsed_data)
-          parsed_data = parsed_data.with_indifferent_access
-          new(parse(parsed_data[:data]).merge :_metadata => parsed_data[:metadata], :_errors => parsed_data[:errors])
+          instantiate_record(self, parsed_data)
+        end
+
+        # Use setter methods of model for each key / value pair in params
+        # Return key / value pairs for which no setter method was defined on the
+        # model
+        #
+        # @private
+        def use_setter_methods(model, params = {})
+          reserved = [:id, model.class.primary_key, *model.class.association_keys]
+          model.class.attributes *params.keys.reject { |k| reserved.include?(k) }
```

Comment:
Hi @zacharywelch , is there any reason to ignore the string keys?

Reply:
Good catch 👍Reserved keys are stored as symbols but attributes can be assigned with string keys. It should instead read 

```ruby
reserved.include?(k.to_sym)
```

Would you like to submit a PR?

Reply:
Thanks for response @zacharywelch . Please help to update it, thanks.